File: solutions/0494-target-sum/solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var findTargetSumWays = function(nums, target) {
    let count = 0;
    
    const memo = new Map();
    
    count = dfs(0,0)
    
    function dfs(sum, i){
        const key = `${sum}:${i}`
        const recall = memo.get(key)
        if(recall !== undefined){
            return recall;
        }
        
        let first = nums[i];
        let subCount = 0
        
        if(i === nums.length - 1){
            if(sum + first === target){
                subCount += 1;
            }
            if(sum - first === target){
                subCount += 1;
            }    
        }else{
            subCount += dfs(sum - first, i + 1);
            subCount += dfs(sum + first, i + 1);    
        }
        memo.set(key, subCount);
        return subCount;
    }
    
    return count;
    
};

File: solutions/0494-target-sum/solution.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        count = 0
        
        size = len(nums)
        
        @cache
        def dfs(sum: int, i: int):
            subcount = 0
            if(i == size -1):
                if sum + nums[i] == target:
                    subcount += 1
                if sum - nums[i] == target:
                    subcount += 1
            else:
                subcount += dfs(sum - nums[i], i+1)
                subcount += dfs(sum + nums[i], i+1)
            
            return subcount
        
        count += dfs(0,0)
        
        return count
